### PR TITLE
Order Email Propagation

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/12448

### CODE REVIEW

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inout/invoicecandidate/M_InOutLine_Handler.java [hunk 1]
```diff
@@ -621,6 +620,10 @@ private void setOrderedData(
 		{
 			icRecord.setC_Order(order);  // also set the order; even if the iol does not directly refer to an order line, it is there because of that order
 			icRecord.setDateOrdered(order.getDateOrdered());
+			if (Check.isEmpty(icRecord.getEMail(), true))
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inout/invoicecandidate/M_InOutLine_Handler.java:623]:
```suggestion
			if (Check.isBlank(icRecord.getEMail()))
```

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/AggregationEngine.java [hunk 2]
```diff
@@ -376,6 +376,7 @@ private void addToInvoiceHeader(
 			invoiceHeader.setC_BPartner_SalesRep_ID(icRecord.getC_BPartner_SalesRep_ID());
 			invoiceHeader.setC_Order_ID(icRecord.getC_Order_ID());
 			invoiceHeader.setPOReference(icRecord.getPOReference()); // task 07978
+			invoiceHeader.setEmail(icRecord.getEMail());
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/AggregationEngine.java:379]: please add "Email" to the `invoicing-agg-std`,  invoice-aggregation rule

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandBLCreateInvoices.java [hunk 3]
```diff
@@ -391,12 +397,21 @@ private I_C_Invoice createInvoice(
 			invoice.setIsSOTrx(header.isSOTrx());
 
 			invoice.setPOReference(invoiceHeader.getPOReference()); // task 07978
+			if(Check.isEmpty(invoice.getEMail(), true))
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandBLCreateInvoices.java:400]:
```suggestion
			if(Check.isBlank(invoice.getEMail()))
```

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandBLCreateInvoices.java [hunk 4]
```diff
@@ -391,12 +397,21 @@ private I_C_Invoice createInvoice(
...
+			}
 			invoice.setC_Order_ID(invoiceHeader.getC_Order_ID()); // set order reference, if any
 			invoice.setC_Async_Batch_ID(invoiceHeader.getC_Async_Batch_ID());
 
 			if (invoiceHeader.getM_InOut_ID() > 0)
 			{
-				invoice.setM_InOut_ID(invoiceHeader.getM_InOut_ID()); // task 06630
+				final I_M_InOut inout = inoutDAO.getById(InOutId.ofRepoId(invoiceHeader.getM_InOut_ID()));
+				invoice.setM_InOut_ID(inout.getM_InOut_ID()); // task 06630
+				if(!Check.isEmpty(inout.getEMail()))
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandBLCreateInvoices.java:411]:
```suggestion
				if(!Check.isBlank(inout.getEMail()))
```

**Review by metas-ts — COMMENTED** [review]
looking at the code I "feel" that we should also have between c) and d):

- c) Take the one from the document's location, if exists
- **⚡ (new) Take the one from the document's user**
- d) Take the one from the document's user's location (if exists)

wdyt?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/AggregationEngine.java (hunk 2 above)
Comment (metas-rc) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/AggregationEngine.java:379]: Done

**Comment by metas-rc** [comment]
> looking at the code I "feel" that we should also have between c) and d):
> 
> * c) Take the one from the document's location, if exists
> * **⚡ (new) Take the one from the document's user**
> * d) Take the one from the document's user's location (if exists)
> 
> wdyt?

The document user is already checked in the second step (b.)

File: backend/de.metas.document.archive/de.metas.document.archive.base/src/main/java/de/metas/document/archive/mailrecipient/impl/OrderDocOutboundLogMailRecipientProvider.java [hunk 5]
```diff
@@ -71,14 +75,28 @@ public String getTableName()
...
+		final String orderEmail = orderRecord.getEMail();
+		final String locationEmail = orderBL.getLocationEmail(OrderId.ofRepoId(orderRecord.getC_Order_ID()));
+
 		final int orderUserRecordId = orderRecord.getAD_User_ID();
 		if (orderUserRecordId > 0)
 		{
 			final DocOutBoundRecipient orderUser = recipientRepository.getById(DocOutBoundRecipientId.ofRepoId(orderUserRecordId));
-			if (!Check.isEmpty(orderUser.getEmailAddress(), true))
+
+			if (!Check.isBlank(orderEmail))
```
Comment (metas-ts) [inline backend/de.metas.document.archive/de.metas.document.archive.base/src/main/java/de/metas/document/archive/mailrecipient/impl/OrderDocOutboundLogMailRecipientProvider.java:86]:
```suggestion
			if (Check.isNotBlank(orderEmail))
```

..for next time :-)

**Review by metas-ts — APPROVED** [review]

**Pull request merged**